core/field_textarea.js

```diff
@@ -0,0 +1,238 @@
+ * @package
+ */
+Blockly.FieldTextArea.prototype.initView = function() {
+  Blockly.FieldDropdown.superClass_.initView.call(this);
+  this.textElement_.setAttribute('class', 'blocklyText blocklyTextCode');
+  
+};
+
+/**
+ * Get the text from this field as displayed on screen.  May differ from getText
+ * due to ellipsis, and other formatting.
+ * @return {string} Currently displayed text.
+ * @private
+ */
+Blockly.FieldTextArea.prototype.getDisplayText_ = function() {
+  var value = this.value_;
+  if (!value) {
+    // Prevent the field from disappearing if empty.
+    return Blockly.Field.NBSP;
+  }
+  value = value.split("\n").map(function(value) {
+    if (value.length > this.maxDisplayLength) {
+      // Truncate displayed string and add an ellipsis ('...').
+      value = value.substring(0, this.maxDisplayLength - 4) + '...';
+    }
+    // Replace whitespace with non-breaking spaces so the value doesn't collapse - keep newline.
+    value = value.replace(/\s/g, Blockly.Field.NBSP);
+    return value;
+  }.bind(this)).join("\n");
+  if (this.sourceBlock_.RTL) {
+    // The SVG is LTR, force value to be RTL.
+    value += '\u200F';
+  }
+  return value;
+};
+
+/**
+ * Updates the text of the textElement.
+ * @protected
+ */
+Blockly.FieldTextArea.prototype.render_ = function() {
+  // Replace the text.
+  var textElement = this.textElement_;
+  // Clear out old tspan lines
+  var currentChild;
+  while (currentChild = textElement.firstChild) {
+    textElement.removeChild(currentChild);
+  }
+  // Add in new tspan lines
+  var txt = this.getDisplayText_();
+  var y = 0;
+  var xoffset = 0;
+  var yoffset = 12.5; // 12.5 is hard-coded in Blockly.Field
+  var txtLines = txt.split("\n");
+  txtLines.forEach(function(t) {
+    Blockly.utils.dom.createSvgElement('tspan', {x:xoffset,y:y + yoffset}, textElement)
+        .appendChild(document.createTextNode(t));
+    y += 20;
+  });
+  if (txtLines.length == 0) {
+    y += 20;
+  }
+
+  // set up widths
+  this.size_.width = this.textElement_.getBBox().width + 5;
+  // Minimum size of block (25) - default borderRect_ height (16)
+  //   = 9 for vertical margin
+  this.size_.height = y + 9;
+
+  if (this.borderRect_) {
+    this.borderRect_.setAttribute('width',
+        this.size_.width + Blockly.BlockSvg.SEP_SPACE_X);
+    this.borderRect_.setAttribute('height', y);
+  }
+};
+
+/**
+ * Create the text input editor widget.
+ * @return {!HTMLInputElement} The newly created text input editor.
+ * @private
+ */
+Blockly.FieldTextArea.prototype.widgetCreate_ = function() {
+  var div = Blockly.WidgetDiv.DIV;
+
+  var htmlInput = document.createElement('textarea');
+  htmlInput.className = 'blocklyHtmlInput';
+  htmlInput.setAttribute('spellcheck', this.spellcheck_);
+  var fontSize =
+      (Blockly.FieldTextInput.FONTSIZE * this.workspace_.scale) + 'pt';
+  div.style.fontSize = fontSize;
+  
+  var scale = this.sourceBlock_.workspace.scale;
+  div.style.lineHeight = (20 * scale) + 'px';
```

Comment:
Ah, I see now. I was dying inside writing out all these magic numbers, since I wasn't sure what fell under "general rendering weirdness" vs not. I'm glad to hear that there's a plan for these.

Reply:
With respect to "general rendering weirdness", we've been rewriting rendering completely, with a lot of success.  The meat of the new rendering code is [here](https://github.com/google/blockly/tree/render/collab/core/block_rendering_rewrite) if you want to check it out.  TLDR is that everything on a field gets an explicit size before we try to render it, and fields are no longer responsible for horizontal and vertical padding.  (And there's a backwards compatibility plan--more on that to come).